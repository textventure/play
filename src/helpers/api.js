/**
 * Gets story from URL.
 *
 * @param  {String} url
 * @return {Promise}
 */
export const getStory = url => {
  return fetch(url)
    .then(response => {
      if (response.ok) {
        return response.text();
      }
      throw new Error('URL: ' + (response.statusText || response.status));
    })
    .then(text => {
      try {
        return window.jsyaml.load(text);
      } catch (error) {
        throw new Error('YAML: Unable to parse');
      }
    })
    .then(story => {
      if (!(story instanceof Object)) {
        throw new Error('YAML: Missing values');
      }
      return story;
    });
};
