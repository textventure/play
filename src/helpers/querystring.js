/**
 * The beginning character of the query string.
 *
 * @constant {String}
 */
const QUESTION_MARK = '?';

/**
 * The substring used to delimit key and value pairs in the query string.
 *
 * @constant {String}
 */
const SEPARATOR = '&';

/**
 * The substring used to delimit keys and values in the query string.
 *
 * @constant {String}
 */
const EQUAL = '=';

/**
 * Parses a URL query string into a collection of key and value pairs.
 *
 * @see {@link https://nodejs.org/api/querystring.html querystring}
 *
 * @param  {String} string - The URL query string to parse.
 * @return {Object}
 */
export const parse = string => {
  if (typeof string !== 'string') {
    throw TypeError('First argument must be a string');
  }

  if (string[0] === QUESTION_MARK) {
    string = string.slice(1);
  }

  if (!string) {
    return {};
  }

  return string.split(SEPARATOR).reduce((accumulator, field) => {
    let [key, value] = field.split(EQUAL);
    key = decodeURIComponent(key);

    if (value === undefined) {
      value = '';
    } else {
      value = decodeURIComponent(value);
    }

    switch (typeof accumulator[key]) {
      // new key-value
      case 'undefined':
        accumulator[key] = value;
        break;

      // existing key-value, convert to array
      case 'string':
        accumulator[key] = [accumulator[key], value];
        break;

      // existing key-values, push to array
      case 'object':
        accumulator[key].push(value);
        break;

      // no default
    }

    return accumulator;
  }, {});
};

export default {
  parse,
};
