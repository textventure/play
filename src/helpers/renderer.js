import React from 'react';

/**
 * Removes newlines from string.
 *
 * @param  {String} string
 * @return {String}
 */
const removeNewlines = string => string.trim().replace(/\n/g, '');

/**
 * Renders text based on format.
 *
 * @param  {String}              input
 * @param  {String}              [format="text"]
 * @return {String|ReactElement}
 */
const render = (input, format = 'text') => {
  switch (format) {
    case 'html':
      return (
        <div dangerouslySetInnerHTML={{ __html: removeNewlines(input) }} />
      );
    case 'markdown':
      return (
        <div
          dangerouslySetInnerHTML={{
            __html: removeNewlines(window.marked(input, { headerIds: false })),
          }}
        />
      );
    case 'text':
    default:
      return input;
  }
};

export default render;
