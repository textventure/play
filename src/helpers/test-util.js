/**
 * `Object.defineProperty` shorthand.
 *
 * @param  {Object} object
 * @param  {String} key
 * @param  {*}      value
 * @return {Object}
 */
export const defineProperty = (object, key, value) => {
  return Object.defineProperty(object, key, {
    value,
    writable: true,
  });
};
