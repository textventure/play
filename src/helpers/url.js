/**
 * The beginning character of the query string.
 *
 * @constant {String}
 */
const QUESTION_MARK = '?';

/**
 * The substring used to delimit key and value pairs in the query string.
 *
 * @constant {String}
 */
const SEPARATOR = '&';

/**
 * The substring used to delimit keys and values in the query string.
 *
 * @constant {String}
 */
const EQUAL = '=';

/**
 * Returns URL search param(s).
 *
 * @see {@link https://developer.mozilla.org/docs/Web/API/URL/searchParams seachParams}
 *
 * @param  {String}        search  - The URL query string.
 * @param  {String}        [param] - The optional parameter name.
 * @return {Object|String}         - The search params or matching param value.
 */
export const searchParams = (search, param) => {
  if (typeof search !== 'string') {
    throw TypeError('First argument must be a string');
  }

  if (search[0] === QUESTION_MARK) {
    search = search.slice(1);
  }

  if (!search) {
    if (typeof param === 'string') {
      return;
    } else {
      return {};
    }
  }

  const searchParamsMap = search
    .split(SEPARATOR)
    .reduce((accumulator, field) => {
      let [key, value] = field.split(EQUAL);
      key = decodeURIComponent(key);

      if (value === undefined) {
        value = '';
      } else {
        value = decodeURIComponent(value);
      }

      switch (typeof accumulator[key]) {
        // new key-value
        case 'undefined':
          accumulator[key] = value;
          break;

        // existing key-value, convert to array
        case 'string':
          accumulator[key] = [accumulator[key], value];
          break;

        // existing key-values, push to array
        case 'object':
          accumulator[key].push(value);
          break;

        // no default
      }

      return accumulator;
    }, {});

  if (typeof param === 'string') {
    return searchParamsMap[param];
  } else {
    return searchParamsMap;
  }
};
