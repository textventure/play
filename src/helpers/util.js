/**
 * Gets key.
 *
 * @param  {String|Object} value
 * @return {String}
 */
export const getKey = value => {
  if (typeof value === 'string') {
    return value;
  }

  // return first property key
  if (value instanceof Object) {
    for (let key in value) {
      return key;
    }
  }
};

/**
 * Gets value.
 *
 * @param  {Object} obj
 * @return {*}
 */
export const getValue = obj => {
  // return first property value
  if (obj instanceof Object) {
    for (let key in obj) {
      return obj[key];
    }
  }
};
