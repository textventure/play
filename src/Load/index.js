import React, { Component } from 'react';
import { Redirect } from 'react-router-dom';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';
import TextField from '@material-ui/core/TextField';
import Typography from '@material-ui/core/Typography';
import Card from '../Card';
import { getStory } from '../helpers/api';

const initialState = {
  error: '',
  value: '',
};
const placeholder = process.env.PUBLIC_URL + '/demo.yaml';

export default class Load extends Component {
  constructor(props) {
    super(props);
    this.state = {
      ...initialState,
      branches: null,
      config: null,
      message: '',
    };

    try {
      const url = new URLSearchParams(props.location.search).get('url');
      if (url) {
        this.state.value = url;
      }
    } catch (error) {}
  }

  /**
   * Handles snackbar close.
   */
  handleClose = () => {
    this.setState({
      message: '',
    });
  };

  /**
   * Handles input change.
   *
   * @param {SyntheticEvent} event
   */
  handleChange = event => {
    const { value } = event.target;
    const newState = {
      ...initialState,
      value,
    };
    try {
      new URL(value);
    } catch (error) {
      newState.error = 'Invalid URL';
    }
    this.setState(newState);
  };

  /**
   * Handles form submit.
   *
   * @param {SyntheticEvent} event
   */
  handleSubmit = event => {
    event.preventDefault();

    getStory(this.state.value)
      .then(story => {
        const { _config: config, ...branches } = story;

        if (config && branches) {
          this.setState({
            branches,
            config,
          });
        }
      })
      .catch(error => {
        this.setState({
          message: error.message,
        });
      });
  };

  render() {
    const { branches, config, value } = this.state;
    if (branches && config) {
      return (
        <Redirect
          to={{
            branches,
            config,
            pathname: '/play',
            search: `url=${encodeURIComponent(value)}`,
          }}
        />
      );
    }

    const { error, message } = this.state;
    return (
      <Card>
        {message && (
          <Snackbar
            autoHideDuration={6000}
            message={message}
            onClose={this.handleClose}
            open={Boolean(message)}
          />
        )}

        <form onSubmit={this.handleSubmit}>
          <Typography gutterBottom variant="headline">
            Load Story
          </Typography>

          <TextField
            autoFocus
            error={Boolean(error)}
            fullWidth
            helperText={error}
            InputLabelProps={{ shrink: true }}
            label="URL"
            margin="normal"
            name="url"
            onChange={this.handleChange}
            placeholder={placeholder}
            required
            spellCheck={false}
            type="url"
            value={value}
          />
          <br />
          <br />

          <Button
            color="primary"
            size="medium"
            type="submit"
            variant="contained"
          >
            Load
          </Button>
        </form>
      </Card>
    );
  }
}
