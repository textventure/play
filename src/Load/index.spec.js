import React from 'react';
import { shallow } from 'enzyme';
import { getStory } from '../helpers/api';
import Load from '.';

jest.mock('../helpers/api', () => ({
  getStory: jest.fn(),
}));

let wrapper;
let props;
let state;
let event;

const { URL, URLSearchParams } = window;

beforeAll(() => {
  window.URL = jest.fn();
  window.URLSearchParams = jest.fn();
});

afterAll(() => {
  window.URL = URL;
  window.URLSearchParams = URLSearchParams;
  jest.unmock('../helpers/api');
});

describe('without props', () => {
  beforeAll(() => {
    wrapper = shallow(<Load />);
  });

  it('renders <TextField> with placeholder', () => {
    expect(wrapper.find('TextField').prop('placeholder')).toBe(
      'http://localhost/example.yaml'
    );
  });

  it('renders correctly', () => {
    expect(wrapper.getElement()).toMatchSnapshot();
  });
});

describe('when props.location.search="?url=http://foo.bar"', () => {
  beforeAll(() => {
    window.URLSearchParams.mockImplementation(search => ({
      get: param => {
        if (param === 'url') {
          return search.replace('?url=', '');
        }
      },
    }));
    props = {
      location: {
        search: '?url=http://foo.bar',
      },
    };
    wrapper = shallow(<Load {...props} />);
  });

  it('sets state.value', () => {
    expect(wrapper.state('value')).toBe(
      props.location.search.replace('?url=', '')
    );
  });
});

describe('with state.branches and state.config', () => {
  beforeAll(() => {
    wrapper = shallow(<Load />);
    state = {
      branches: {},
      config: {},
    };
    wrapper.setState(state);
  });

  it('renders <Redirect>', () => {
    expect(wrapper.find('Redirect').length).toBe(1);
  });

  it('redirects to "/play"', () => {
    expect(wrapper.find('Redirect').prop('to')).toEqual({
      ...state,
      pathname: '/play',
      search: 'url=',
    });
  });

  describe('and state.value="http://foo.bar"', () => {
    beforeAll(() => {
      wrapper.setState({
        value: 'http://foo.bar',
      });
    });

    it('renders correctly', () => {
      expect(wrapper.getElement()).toMatchSnapshot();
    });
  });
});

describe('when state.message="Message"', () => {
  beforeAll(() => {
    wrapper = shallow(<Load />);
    state = {
      message: 'Message',
    };
    wrapper.setState(state);
  });

  it('renders <Snackbar> with message', () => {
    expect(wrapper.find('WithStyles(Snackbar)').prop('message')).toBe(
      state.message
    );
  });

  it('renders correctly', () => {
    expect(wrapper.getElement()).toMatchSnapshot();
  });

  it('sets state.message="" when `handleClose` is called', () => {
    wrapper.instance().handleClose();
    expect(wrapper.state('message')).toBe('');
  });
});

describe('when state.value="http://foo.bar"', () => {
  beforeAll(() => {
    wrapper = shallow(<Load />);
    state = {
      value: 'http://foo.bar',
    };
    wrapper.setState(state);
  });

  it('renders <TextField> with value', () => {
    expect(wrapper.find('TextField').prop('value')).toBe(state.value);
  });
});

describe('when state.error="Error"', () => {
  beforeAll(() => {
    wrapper = shallow(<Load />);
    wrapper.setState({
      error: 'Error',
    });
  });

  it('renders <TextField> with error styling', () => {
    expect(wrapper.find('TextField').prop('error')).toBe(true);
  });
});

describe('onChange', () => {
  beforeAll(() => {
    wrapper = shallow(<Load />);
  });

  describe('when value is a url', () => {
    beforeAll(() => {
      event = {
        target: {
          value: 'http://foo.bar',
        },
      };
      wrapper.find('TextField').simulate('change', event);
    });

    it('sets state.value', () => {
      expect(wrapper.state('value')).toBe(event.target.value);
    });

    it('does not set state.error', () => {
      expect(wrapper.state('error')).toBe('');
    });
  });

  describe('when value is not a url', () => {
    beforeAll(() => {
      window.URL.mockImplementation(() => {
        throw new Error();
      });
      event = {
        target: {
          value: 'foo.bar',
        },
      };
      wrapper.find('TextField').simulate('change', event);
    });

    it('sets state.value', () => {
      expect(wrapper.state('value')).toBe(event.target.value);
    });

    it('sets state.error', () => {
      expect(wrapper.state('error')).toBe('Invalid URL');
    });
  });
});

describe('onSubmit', () => {
  let resolvedValue;
  let rejectedValue;

  describe('when successful', () => {
    beforeAll(() => {
      resolvedValue = {
        _config: {},
        branches: {},
      };
      getStory.mockImplementation(
        () => new Promise(resolve => resolve(resolvedValue))
      );

      wrapper = shallow(<Load />);
      state = {
        value: 'http://foo.bar',
      };
      wrapper.setState(state);

      event = {
        preventDefault: jest.fn(),
      };
      wrapper.find('form').simulate('submit', event);
    });

    it('prevents default event', () => {
      expect(event.preventDefault).toHaveBeenCalled();
      event.preventDefault.mockReset();
    });

    it('calls `getStory`', () => {
      expect(getStory).toHaveBeenCalledWith(state.value);
    });

    it('sets state with branches and config', () => {
      const { _config: config, ...branches } = resolvedValue;
      expect(wrapper.state('branches')).toEqual(branches);
      expect(wrapper.state('config')).toEqual(config);
    });

    it('redirects to "/play"', () => {
      wrapper.update();
      const { _config: config, ...branches } = resolvedValue;

      expect(wrapper.find('Redirect').prop('to')).toEqual({
        branches,
        config,
        pathname: '/play',
        search: `url=${encodeURIComponent(state.value)}`,
      });
    });
  });

  describe('with missing properties', () => {
    beforeAll(() => {
      resolvedValue = {};
      getStory.mockImplementation(
        () => new Promise(resolve => resolve(resolvedValue))
      );

      wrapper = shallow(<Load />);
      state = {
        value: 'http://foo.bar',
      };
      wrapper.setState(state);
      wrapper.instance().handleSubmit({ preventDefault: () => {} });
    });

    it('does not set state', () => {
      expect(wrapper.state('branches')).toEqual(null);
      expect(wrapper.state('config')).toEqual(null);
    });

    it('does not redirect to "/play"', () => {
      wrapper.update();
      expect(wrapper.find('Redirect').length).toBe(0);
    });
  });

  describe('when unsuccessful', () => {
    beforeAll(() => {
      rejectedValue = {
        message: 'Error',
      };
      getStory.mockImplementation(
        () => new Promise((resolve, reject) => reject(rejectedValue))
      );

      wrapper = shallow(<Load />);
      state = {
        value: 'http://foo.bar',
      };
      wrapper.setState(state);

      event = {
        preventDefault: jest.fn(),
      };
      wrapper.find('form').simulate('submit', event);
    });

    it('prevents default event', () => {
      expect(event.preventDefault).toHaveBeenCalled();
      event.preventDefault.mockReset();
    });

    it('calls `getStory`', () => {
      expect(getStory).toHaveBeenCalledWith(state.value);
    });

    it('sets state.message', () => {
      expect(wrapper.state('message')).toBe(rejectedValue.message);
    });

    it('renders <Snackbar> with message', () => {
      wrapper.update();
      expect(wrapper.find('WithStyles(Snackbar)').prop('message')).toBe(
        rejectedValue.message
      );
    });
  });
});
