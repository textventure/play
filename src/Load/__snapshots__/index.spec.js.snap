// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when props={} renders correctly 1`] = `
<ForwardRef(WithStyles)>
  
  <form
    onSubmit={[Function]}
  >
    <ForwardRef(WithStyles)
      gutterBottom={true}
      variant="h5"
    >
      Load Story
    </ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      autoFocus={true}
      error={false}
      fullWidth={true}
      helperText=""
      label="URL"
      margin="normal"
      name="url"
      onChange={[Function]}
      placeholder="http://localhost/demo.yaml"
      required={true}
      spellCheck={false}
      type="url"
      value=""
    />
    <br />
    <br />
    <ForwardRef(WithStyles)
      color="primary"
      size="medium"
      type="submit"
      variant="contained"
    >
      Load
    </ForwardRef(WithStyles)>
  </form>
</ForwardRef(WithStyles)>
`;

exports[`when state.message="foo" renders correctly 1`] = `
<ForwardRef(WithStyles)>
  <ForwardRef(WithStyles)
    autoHideDuration={6000}
    message="foo"
    onClose={[Function]}
    open={true}
  />
  <form
    onSubmit={[Function]}
  >
    <ForwardRef(WithStyles)
      gutterBottom={true}
      variant="h5"
    >
      Load Story
    </ForwardRef(WithStyles)>
    <ForwardRef(WithStyles)
      InputLabelProps={
        Object {
          "shrink": true,
        }
      }
      autoFocus={true}
      error={false}
      fullWidth={true}
      helperText=""
      label="URL"
      margin="normal"
      name="url"
      onChange={[Function]}
      placeholder="http://localhost/demo.yaml"
      required={true}
      spellCheck={false}
      type="url"
      value=""
    />
    <br />
    <br />
    <ForwardRef(WithStyles)
      color="primary"
      size="medium"
      type="submit"
      variant="contained"
    >
      Load
    </ForwardRef(WithStyles)>
  </form>
</ForwardRef(WithStyles)>
`;
